tsc --watch does not correctly account for added & deleted files

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.1.1

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** tsc --watch

**Code**

```ts
// a.ts
declare module 'a' {
  type foo = number;
}

// tsconfig.json
{}
```

Run `tsc --watch --noEmit --project .`.  Wait for typechecking to complete.  `cp a.ts b.ts`; wait for typechecking to complete again.  then `touch *` and wait for typechecking to run again.  then `rm b.ts` and wait for typechecking to complete again

**Expected behavior:**

Typechecking should succeed on the first run, should fail with 2 errors on the second run (`error TS2300: Duplicate identifier 'foo'`) since a.ts and b.ts duplicate the same module; the third run should come up with the same results, and the 4th run should succeed.

**Actual behavior:**

The first run passes as expected, the second run only detects the error in `b.ts`, the touch * causes the 3rd run to recognize both errors, and then deleting b.ts causes the 4th run to still think there's an error in `a.ts`.

If I had to guess, tsc --watch is assuming in these file add / remove cases that adding or removing b.ts isn't going to change the type errors it finds in a.ts; this type of error is an example where that's clearly not the case.

**Related Issues:** couldn't find any

--- a/src/compiler/builderState.ts
+++ b/src/compiler/builderState.ts
@@ -148,7 +148,38 @@ namespace ts.BuilderState {
             });
         }
 
+        // Add module augmentation as references
+        if (sourceFile.moduleAugmentations.length) {
+            const checker = program.getTypeChecker();
+            for (const moduleName of sourceFile.moduleAugmentations) {
+                if (!isStringLiteral(moduleName)) { continue; }
+                const symbol = checker.getSymbolAtLocation(moduleName);
+                if (!symbol) { continue; }
+
+                // Add any file other than our own as reference
+                addReferenceFromAmbientModule(symbol);
+            }
+        }
+
+        // From ambient modules
+        for (const ambientModule of program.getTypeChecker().getAmbientModules()) {
+            if (ambientModule.declarations.length > 1) {
+                addReferenceFromAmbientModule(ambientModule);
+            }
+        }
+
         return referencedFiles;
+
+        function addReferenceFromAmbientModule(symbol: Symbol) {
+            // Add any file other than our own as reference
+            for (const declaration of symbol.declarations) {
+                const declarationSourceFile = getSourceFileOfNode(declaration);
+                if (declarationSourceFile &&
+                    declarationSourceFile !== sourceFile) {
+                    addReferencedFile(declarationSourceFile.resolvedPath);
+                }
+            }
+        }
 
         function addReferencedFile(referencedPath: Path) {
             if (!referencedFiles) {

--- a/src/testRunner/unittests/tscWatchMode.ts
+++ b/src/testRunner/unittests/tscWatchMode.ts
@@ -1386,6 +1386,41 @@ foo().hello`
             // File a need not be rewritten
             assert.equal(host.getModifiedTime(`${currentDirectory}/a.js`), modifiedTimeOfAJs);
         });
+
+        it("updates errors when ambient modules of program changes", () => {
+            const currentDirectory = "/user/username/projects/myproject";
+            const aFile: File = {
+                path: `${currentDirectory}/a.ts`,
+                content: `declare module 'a' {
+  type foo = number;
+}`
+            };
+            const config: File = {
+                path: `${currentDirectory}/tsconfig.json`,
+                content: "{}"
+            };
+            const files = [aFile, config, libFile];
+            const host = createWatchedSystem(files, { currentDirectory });
+            const watch = createWatchOfConfigFile("tsconfig.json", host);
+            checkProgramActualFiles(watch(), [aFile.path, libFile.path]);
+            checkOutputErrorsInitial(host, emptyArray);
+
+            // Create bts with same file contents
+            const bTsPath = `${currentDirectory}/b.ts`;
+            host.writeFile(bTsPath, aFile.content);
+            host.runQueuedTimeoutCallbacks();
+            checkProgramActualFiles(watch(), [aFile.path, "b.ts", libFile.path]);
+            checkOutputErrorsIncremental(host, [
+                "a.ts(2,8): error TS2300: Duplicate identifier 'foo'.\n",
+                "b.ts(2,8): error TS2300: Duplicate identifier 'foo'.\n"
+            ]);
+
+            // Delete bTs
+            host.deleteFile(bTsPath);
+            host.runQueuedTimeoutCallbacks();
+            checkProgramActualFiles(watch(), [aFile.path, libFile.path]);
+            checkOutputErrorsIncremental(host, emptyArray);
+        });
     });
 
     describe("tsc-watch emit with outFile or out setting", () => {

